// @sanity/runtime-cli/dist/commands/functions/invoke.js
import { Args, Command, Flags } from '@oclif/core';
import readBlueprintOnDisk from '../../actions/blueprints/read-blueprint.js';
import { invoke } from '../../actions/functions/invoke.js';
import config from '../../config.js';
import { red } from '../../utils/display/colors.js';
import Spinner from '../../utils/spinner.js';
export default class Invoke extends Command {
    static args = {
        id: Args.string({ description: 'The ID of the function to invoke', required: true }),
    };
    static description = 'Invoke a remote Sanity Function';
    static examples = [
        `<%= config.bin %> <%= command.id %> <ID> --data '{ "id": 1 }'`,
        `<%= config.bin %> <%= command.id %> <ID> --file 'payload.json'`,
    ];
    static flags = {
        data: Flags.string({ char: 'd', description: 'Data to send to the function', required: false }),
        file: Flags.string({
            char: 'f',
            description: 'Read data from file and send to the function',
            required: false,
        }),
    };
    async run() {
        const { args, flags } = await this.parse(Invoke);
        const s = new Spinner();
        const { deployedStack } = await readBlueprintOnDisk({ getStack: true });
        if (!deployedStack)
            this.error('Stack not found'); // returns
        const { projectId } = deployedStack;
        s.start(`Invoking function "${args.id}"`);
        const result = await invoke(args.id, { data: flags.data, file: flags.file }, { token: config.token, projectId });
        if (result.ok) {
            s.stop(`Invocation of ${args.id} succeeded`);
            this.log(JSON.stringify(result.json, null, 2));
        }
        else {
            s.stop(`${red('Failed')} to invoke function`);
            this.log(`Error: ${result.error || 'Unknown error'}`);
        }
    }
}
